base/iobuffer.jl

```diff
@@ -1,26 +1,117 @@
-    end
+    # If not seekable, implementation is free to destroy (compact) data in 1:mark-1.
+    # If it IS seekable, the user may always recover any data in 1:size by seeking,
+    # so no data can be destroyed.
+    # Non-seekable IOBuffers can only be constructed with `PipeBuffer`, which are writable,
+    # readable and append.
+    seekable::Bool
+
+    # If true, write new data to the index size+1 instead of the index ptr.
+    append::Bool
+
+    # Last index of `data` that has been written to. Data in size+1:end has not yet been used,
+    # and may contain arbitrary values.
+    # This value is always in 0 : lastindex(data)
+    size::Int
+
+    # When the buffer is resized, or a new buffer allocated, this is the maximum size of the buffer.
+    # A new GenericIOBuffer may be constructed with an existing data larger than `maxsize`.
+    # When that happens, the buffer will not write to data in maxsize + 1 : lastindex(data).
+    # This value is always in 0:typemax(Int).
+    maxsize::Int
+
+    # Data is read/written from/to ptr, except in situations where append is true, in which case
+    # data is still read from ptr, but written to size+1.
+    # This value is always in offset + 1 : size+1
+    ptr::Int
+
+    # This is used when seeking. seek(io, 0) results in ptr == offset.
+    # The offset is needed because, if a buffer is instantiated from a Vector with a non-zero
+    # memory offset, the start of the vector, and thus the start of data, does not correspond
+    # to the start of its underlying memory.
+    # Once the offset is set to zero, it will never be set to nonzero.
+    # This is always in 0:lastindex(data)
+    offset::Int
+
+    # mark is the position (as given by `position`, i.e. io.ptr - io.offset - 1)
+    # which can be seeked back using `reset`, even for non-seekable buffers.
+    # For non-seekable buffers that can be compacted, data before the mark can be
+    # destroyed.
+    # This value is always in -1 : size - offset
+    mark::Int
+
+    # Unsafe constructor which does not do any checking
+    global function _new_generic_iobuffer(
+            ::Type{T},
+            data::T,
+            readable::Bool,
+            writable::Bool,
+            seekable::Bool,
+            append::Bool,
+            maxsize::Int,
+        ) where T<:AbstractVector{UInt8}
+        len = Int(length(data))::Int
+        return new{T}(data, false, readable, writable, seekable, append, len, maxsize, 1, 0, -1)
+    end
+end
+
+function GenericIOBuffer{T}(
+        data::T,
+        readable::Bool,
+        writable::Bool,
+        seekable::Bool,
+        append::Bool,
+        maxsize::Integer,
+    ) where T<:AbstractVector{UInt8}
+    require_one_based_indexing(data)
+    mz = Int(maxsize)::Int
+    len = Int(length(data))::Int
+    if mz < len
+        throw(ArgumentError("maxsize must not be smaller than data length"))
+    end
```

Comment:
This error seems a bit arbitrary, since as the comments elsewhere note, the implementation has to support it anyways and it has a fairly well-defined meaning (restricts resizing), and adding the error complicates the constructor situation (with `_new_generic_iobuffer`). Maybe we also instead need to pass the length directly as an argument to this constructor instead of inferring intent from `length(data)`? Anyways, I guess this is just a nit, and not actually important to change.

Reply:
The implementation in this PR only supports `length(data) > maxsize` in the case where a `Vector` is passed in, where the legnth of the vector does not exceed `maxsize`, but the underlying memory does. When that is the case, we know that the user never has more than `maxsize` bytes of used data in the buffer. If we ever have that, it becomes a problem, because if we then need to resize (or reallocate the data) for some reason, we discard user data.